### PR TITLE
GH-36317: [C++] Return a BufferVector from CleanListOffsets

### LINKED ISSUE #36317: [C++] Simplify CleanListOffsets
### Describe the enhancement requested

It takes output parameters, but it could return a `Result<BufferVector>` instead.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

`CleanListOffsets` is taking output parameters when it could return multiple buffers in a vector exactly how they are supposed to be used to construct the list-like array.

Besides that, having the logic for setting the validity buffer split between it and the caller is unnecessarily confusing.

### What changes are included in this PR?

 - Addition of a new constructor for `MapArray` that can accept a pre-allocated `BufferVector`
 - Removal of output parameter in `CleanListOffsets` and change of return type

### Are these changes tested?

Yes, by existing tests. A pre-existing `MapArray` ctor delegates to this new ctor, so all tests that exercise the old ctor also exercise the new.
* Closes: #36317

### CODE REVIEW

**Comment by kou** [comment]
Could you open an issue for non MINOR change?

Our MINOR definition: https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes
(This is also linked from the auto generated message: https://github.com/apache/arrow/pull/36316#issuecomment-1608586156 )

**Comment by felipecrv** [comment]
> Could you open an issue for non MINOR change?
> 
> Our MINOR definition: https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes (This is also linked from the auto generated message: [#36316 (comment)](https://github.com/apache/arrow/pull/36316#issuecomment-1608586156) )

Done.

File: cpp/src/arrow/array/array_nested.cc [hunk 1]
```diff
@@ -74,9 +76,8 @@ Status CleanListOffsets(const Array& offsets, MemoryPool* pool,
     // Copy valid bits, ignoring the final offset (since for a length N list array,
     // we have N + 1 offsets)
     ARROW_ASSIGN_OR_RAISE(
-        auto clean_valid_bits,
+        auto clean_validitty_buffer,
```
Comment (pitrou) [inline cpp/src/arrow/array/array_nested.cc:79]: Obvious typo in variable name here.

**Review by pitrou — CHANGES_REQUESTED** [review]
LGTM except for one nit. Thanks @felipecrv !

File: cpp/src/arrow/array/array_nested.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/array/array_nested.cc]: Fixed.

**Review by pitrou — APPROVED** [review]

**Pull request merged**